base/thread_pool.cc

```diff
@@ -26,40 +28,55 @@ thread_pool_impl::~thread_pool_impl() {
   }
 }
 
-bool thread_pool_impl::dequeue(task& t, std::vector<thread_pool_impl::task_id>& task_stack) {
+namespace {
+
+thread_local std::vector<thread_pool::task_id> task_stack;
+
+}  // namespace
+
+thread_pool::task_id thread_pool_impl::dequeue(task& t) {
   for (auto i = task_queue_.begin(); i != task_queue_.end(); ++i) {
-    if (std::find(task_stack.begin(), task_stack.end(), std::get<2>(*i)) != task_stack.end()) {
+    const task_id id = std::get<2>(*i);
```

Comment:
style nit: would be easier to read if you used local aliases for std::get<0> and std::get<1> as well

Reply:
Done for `std::get<0>`, but `std::get<1>` is already immediately assigned to the relevant variable.